man/kmrshell.1
'\"
.\" Copyright (C) 2012-2018 RIKEN R-CCS
.\" nroff -u0 -Tlp -man
.TH kmrshell 1 "2014-02-04" "KMR" "User Commands"
.SH NAME
kmrshell \- map-reduce by shell command pipeline
.SH SYNOPSIS
.LP
.nf
\fBkmrshell\fR -m mapper -r reducer file
.fi
.SH DESCRIPTION
.sp
.LP
kmrshell performs map-reduce by pipelining shell commands, also called
"streaming" in Hadoop.  It forks and execs processes of a mapper, a
shuffler, and a reducer, and they are joined through pipes.  kmrshell
reads a file and passes the contents to a mapper.  When a directory
name is given instead of a file, kmrshell reads all regular files
under the directory and passes them to a mapper (not recursive).
Shuffler is written with KMR.
.SH OPTIONS
.sp
.LP
The following options are supported:
.sp
.ne 2
.na
-m mapper
.ad
.sp .6
.RS 4n
Specifies a mapper program.  The program can have arguments, where
they are separated by a whitespace.
.sp
Mapper specification: A mapper reads data from STDIN and outputs
key-value data to STDOUT.  Output data is a sequence of a line "key
value\\n", where the fields are separated by a whitespace.
.RE
.sp
.ne 2
.na
-r reducer
.ad
.sp .6
.RS 4n
Specifies a reducer program.  The program can have arguments, where
they are separated by a whitespace.
.sp
Reducer Specification: A reducer reads key-value data from STDIN and
outputs the result to STDOUT.  Input data is a sequence of a line "key
value\\n", where the fields are separated by a whitespace.
Lines with the same keys constitutes consecutive lines.
.RE


CHANGELOG.md
# CHANGELOG

The list is sketchy.

## 1.9 (2018-08-27)

- 1.9 is a minor update
- Move GitHub repository from pf-aics-riken to riken-rccs
- Change the copyright holder name: RIKEN AICS to RIKEN R-CCS

## 1.8.1 (2016-04-25)

- Add ex/kmreprun.py, an example of simple shell command invoker
- Improve support of Mac OS X
- Several small bug fixes

## 1.8.0 (2015-12-25)

- Add a function tracing option for KMRViz, a KMR visualizer
- Add a simple scan function, kmr_scan_on_value()
- Add a load-leveling shuffle, kmr_shuffle_leveling_pair_count()
- Add a new kmr example, pagerank.c
- BUGFIX: kmr_match(), kmr_histogram_count_by_ranks(), kmr_find_key()

## 1.7.1 (2015-06-29)

- BUGFIX: kmr_take_one()
- BUGFIX: kmrrungenscript.py

## 1.7 (2015-06-22)

- Change license to BSD
- Strictly check options passed to kmr functions
- Add support for FOCUS supercomputer in kmrrungenscript.py
- Add Python API

## 1.6 (2015-04-01)

- Disable keep_open option in kmr functions except mapper functions
- Support no-fsync mode in Checkpoint/Restart
- Improve performance of KMRRUN when many input files are given
- BUGFIX: Checkpoint/Restart
- EXPERIMENTAL: Support selective mode that reduces IO overhead in
  Checkpoint/Restart

## 1.5 (2015-02-04)

- Reduce overhead of KMRRUN
- Temporal files for reducers are generated in subdirectories when
  using KMRRUN
- WORKAROUND: Abort on calling kmr_read_file_xxx with some MPI
  implementations
- WORKAROUND: Abort on calling kmr_map_processes and using KMRRUN
  with some Open MPI versions (1.6.3 - 1.8.1)

## 1.4 (2014-12-25)

- Resolved a scalability issue of KMRRUN and map functions that
  spawns child programs

## 1.3.2 (2014-11-07)

- Improve job script generator for KMRRUN

## 1.3.1 (2014-11-04)

- Add a word counting sample of KMRRUN
- BUGFIX: KMRRUN command line option parsing

## 1.3 (2014-09-17)

- Add a new function for affinity-aware file assignment

## 1.2 (2014-06-27)

- Support Checkpoint/Restart
- Add KMRRUN command that executes simple MapReduce workflow

## 1.1 (2013-09-20)

- 1.1 is a minor update

## 1.0 (2013-04-26)

- First release


shell/kmrshell.c
/* kmrshell.c */
/* Copyright (C) 2012-2018 RIKEN R-CCS */

/** \file kmrshell.c KMR-Shell for Streaming.  It forks processes of a
    mapper, a shuffler, and a reducer, then, it reads a number of
    specified input files and passes their data to a mapper via a
    pipe.  A mapper, a shuffler, and a reducer are shell
    executables.  */

#include <stdio.h>
#include <stdlib.h>
#include <stddef.h>
#include <string.h>
#include <unistd.h>
#include <fcntl.h>
#include <signal.h>
#include <sys/types.h>
#include <sys/stat.h>
#include <sys/wait.h>
#include <sys/param.h>
#include <dirent.h>
#include <assert.h>

/** Maximum length of a line of data. */
#define LINELEN 32767

/** Maximum number of arguments to mapper and reducer programs. */
#define ARGSIZ 8

/** Buffer string size of arguments to mapper and reducer programs. */
#define ARGSTRLEN (8 * 1024)

/** Parameter for pipeops(). */
enum {
    PIPE_IN = 1,	/**< Attach stdin to pipe. */
    PIPE_OUT = 2,	/**< Attach stdout to pipe. */
};

static char *shuffler = "kmrshuffler";

static int pipeops(int *, int);
static int execute(char *, char *const [], pid_t *);
static void reapchild(int);
static void putfilecontents(char *);
static int filereader(char *);
static void parse_args(char *, char *[]);

/** Handles SIGCHLD signal. */

static void
reapchild(int exitstat)
{
    int status;
    //fprintf(stderr, "reapchild\n");
    waitpid(-1, &status, 0);
}

/** Sets up pipe states.  It attaches pipes to stdin or stdout.  */

static int
pipeops(int *pipefd, int direc)
{
    int ret;
    if (direc == PIPE_IN) {
        if ((ret = close(pipefd[1])) < 0) {
	    perror("close pipe");
	    return ret;
	}
	if ((ret = dup2(pipefd[0], STDIN_FILENO)) < 0) {
	    perror("dup2 pipe");
	    return ret;
	}
	if ((ret = close(pipefd[0])) < 0) {
	    perror("close pipe");
	    return ret;
	}
    }

    if (direc == PIPE_OUT) {
        if ((ret = close(pipefd[0])) < 0) {
	    perror("close pipe");
	    return ret;
	}
	if ((ret = dup2(pipefd[1], STDOUT_FILENO)) < 0) {
	    perror("dup2 pipe");
	    return ret;
	}
	if ((ret = close(pipefd[1])) < 0) {
	    perror("close pipe");
	    return ret;
	}
    }
    return 0;
}

/** Forks and execs a process.
    \param prog program path.
    \param args argments to program.
    \param cldpid pid of child (out). */

static int
execute(char *prog, char *const args[], pid_t *cldpid)
{
    pid_t pid;
    int ret, pipefd[2];
    struct sigaction schld, spipe;

    memset(&schld, 0, sizeof(struct sigaction));
    schld.sa_handler = reapchild;
    schld.sa_flags |= SA_RESTART;
    ret = sigaction(SIGCHLD, &schld, NULL);
    assert(ret >= 0);

    memset(&spipe, 0, sizeof(struct sigaction));
    spipe.sa_handler = SIG_IGN;
    ret = sigaction(SIGPIPE, &spipe, NULL);
    assert(ret >= 0);

    if ((ret = pipe(pipefd)) < 0) {
        perror("pipe");
	return ret;
    }

    pid = fork();

    if (pid < 0) {
        perror("fork");
	return -1;
    }

    if (pid > 0) {
        /* parent */
        if ((ret = pipeops(pipefd, PIPE_OUT)) < 0) {
	    return -1;
	}
	if (cldpid != NULL) {
	    *cldpid = pid;
	}
    }

    if (pid == 0) {
        /* child */
        if ((ret = pipeops(pipefd, PIPE_IN)) < 0) {
	    return ret;
	}
	if ((ret = execv(prog, args)) < 0) {
	    return ret;
	}

    }
    return 0;
}

/* Puts contents of a file to stdout.  It reads a line from a file and
   put it.
   \param path to file. */

static void
putfilecontents(char *path)
{
    FILE *fin;
    char line[LINELEN];

    fin = fopen(path, "r");
    if (fin == NULL) {
	perror("open");
	return;
    }

    while ((fgets(line, sizeof(line), fin)) != NULL) {
	if (fputs(line, stdout) == EOF) {
	    break;
	}
    }

    fclose(fin);
    return;
}

/** Reads files (file-reader).  It reads possibly multiple files and
    writes their contents to stdout.  If "path" is a directory, it
    enumerates the files under the directory and reads each file.
    \param path path to file (or directory). */

static int
filereader(char *path)
{
    struct stat status;

    /* file check */
    if (stat(path, &status) < 0) {
        fprintf(stderr, "file %s error\n", path);
        return -1;
    }

    if (!S_ISDIR(status.st_mode) && !S_ISREG(status.st_mode)) {
        fprintf(stderr, "file %s is not regular or directory\n", path);
        return -1;
    }

    if (S_ISDIR(status.st_mode)) {
	/* directory */
	/*char b[MAXGETDENTS_SIZE];*/
	size_t direntsz;
	long nmax = pathconf(path, _PC_NAME_MAX);
	if (nmax == -1) {
	    direntsz = (64 * 1024);
	} else {
	    direntsz = (offsetof(struct dirent, d_name) + (size_t)nmax + 1);
	}
	DIR *d;
	struct dirent *dentp;
	char b[direntsz];

	d = opendir(path);
	if (d == NULL) {
	    perror("opendir");
	    return -1;
	}
	while (readdir_r(d, (void *)b, &dentp) >= 0) {
	    struct stat substat;
	    char fullpath[MAXPATHLEN];
	    int ret;
	    if (dentp == NULL) {
		break;
	    }

	    ret = snprintf(fullpath, sizeof(fullpath), "%s/%s", path, dentp->d_name);
	    if (ret <= 0) {
		perror("snprintf");
		continue;
	    }

	    if (stat(fullpath, &substat) < 0) {
		continue;
	    }

	    if (S_ISREG(substat.st_mode)) {
		putfilecontents(fullpath);
	    }
	}
	closedir(d);
    } else {
	putfilecontents(path);
    }
    return 0;
}

/** Parses command parameters given for mapper and reducer arguments.
    It scans an argument string like "mapper arg0 arg1" for the -m and
    -r options, and generates an argv array {"mapper", "arg0", "arg1",
    0}.  The separator is a whitespace.
    \param argstr string given for -m or -r options.
    \param argary array to be filled by argument strings. */

static void
parse_args(char *argstr, char *argary[])
{
    char *cp, *np;
    char **ap;

    ap = argary;
    cp = argstr;
    while (1) {
	*ap = cp;
	if ((np = strchr((const char*)cp, ' ')) != NULL) {
	    *np++ = '\0';
	}
	if (++ap >= &argary[ARGSIZ-1]) {
	    **ap = '\0';
	    break;
	} else {
	    if (np == NULL) {
		**ap = '\0';
		break;
	    }
	}
	while (*np == ' ') {
	    np++;
	}
	cp = np;
    }
}

/** Starts map-reduce shell processes (for "streaming").  It
    forks and execs a mapper, a shuffler, and a reducer, which are
    connected via pipes.
    \arg \c -m \c mapper program.
    \arg \c -r \c reducer program.
    \arg \c input file or directory. */

int
main(int argc, char *argv[])
{
    int ret, opt;
    char *mapper = NULL, *reducer = NULL, *infile = NULL;
    char *margv[ARGSIZ], *rargv[ARGSIZ], *sargv[2];
    pid_t mapper_pid, shuffler_pid, reducer_pid;
    char margbuf[ARGSTRLEN];
    char rargbuf[ARGSTRLEN];

    while ((opt = getopt(argc, argv, "m:r:")) != -1) {
	switch (opt) {
	    size_t asz;
	case 'm':
	    asz = (strlen(optarg) + 1);
	    if (asz >= ARGSTRLEN) {
		fprintf(stderr, "Argument too long for mapper (%s)\n",
			optarg);
		exit(-1);
	    }
	    memcpy(margbuf, optarg, asz);
	    parse_args(margbuf, &margv[0]);
	    mapper = margv[0];
	    break;
	case 'r':
	    asz = (strlen(optarg) + 1);
	    if (asz >= ARGSTRLEN) {
		fprintf(stderr, "Argument too long for reducer (%s)\n",
			optarg);
		exit(-1);
	    }
	    memcpy(rargbuf, optarg, asz);
	    parse_args(rargbuf, &rargv[0]);
	    reducer = rargv[0];
	    break;
	default:
	    fprintf(stderr, "Usage %s -m mapper [-r reducer] inputfile\n", argv[0]);
	    exit(-1);
	}
    }

    if ((argc - optind) != 1) {
	fprintf(stderr, "Usage %s -m mapper [-r reducer] inputfile\n", argv[0]);
	exit(EXIT_FAILURE);
    } else {
	infile = argv[optind];
	optind++;
    }

    if (mapper == NULL) {
	fprintf(stderr, "Usage %s -m mapper [-r reducer] inputfile\n", argv[0]);
	exit(EXIT_FAILURE);
    }

    if (reducer != NULL) {
        /* Start reducer process. */
        ret = execute(reducer, rargv, &reducer_pid);
	if (ret < 0) {
	    fprintf(stderr, "execute %s failed\n", reducer);
	    exit(EXIT_FAILURE);
	}

	/* Start shuffler process. */
	sargv[0] = shuffler;
	sargv[1] = NULL;
	ret = execute(shuffler, sargv, &shuffler_pid);
	if (ret < 0) {
	    fprintf(stderr, "execute %s failed\n", shuffler);
	    exit(EXIT_FAILURE);
	}
    }

    /* Start mapper process. */
    ret = execute(mapper, margv, &mapper_pid);
    if (ret < 0) {
	fprintf(stderr, "execute %s failed\n", mapper);
	exit(EXIT_FAILURE);
    }

    /* Call filereader(). */
    ret = filereader(infile);
    if (ret < 0) {
	fprintf(stderr, "filereader failed\n");
    }
    fclose(stdout);

    if (reducer != NULL) {
        /* Wait for termination of shuffler. */
        waitpid(reducer_pid, NULL, 0);
    } else {
        /* Wait for termination of mapper. */
        waitpid(mapper_pid, NULL, 0);
    }

    exit(EXIT_SUCCESS);
}

/*
Copyright (C) 2012-2018 RIKEN R-CCS
This library is distributed WITHOUT ANY WARRANTY.  This library can be
redistributed and/or modified under the terms of the BSD 2-Clause License.
*/
